activerecord/lib/active_record/relation/batches.rb

```diff
@@ -251,25 +251,31 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
           end
         end
 
-        attr = Relation::QueryAttribute.new(primary_key, primary_key_offset, klass.type_for_attribute(primary_key))
-        batch_relation = relation.where(arel_attribute(primary_key).gt(Arel::Nodes::BindParam.new(attr)))
+        bind = bind_attribute(primary_key_offset)
+        batch_relation = relation.where(arel_attribute(primary_key).gt(bind))
       end
     end
 
     private
 
       def apply_limits(relation, start, finish)
-        if start
-          attr = Relation::QueryAttribute.new(primary_key, start, klass.type_for_attribute(primary_key))
-          relation = relation.where(arel_attribute(primary_key).gteq(Arel::Nodes::BindParam.new(attr)))
-        end
-        if finish
-          attr = Relation::QueryAttribute.new(primary_key, finish, klass.type_for_attribute(primary_key))
-          relation = relation.where(arel_attribute(primary_key).lteq(Arel::Nodes::BindParam.new(attr)))
-        end
+        relation = apply_start_limit(relation, start) if start
+        relation = apply_finish_limit(relation, finish) if finish
         relation
       end
 
+      def apply_start_limit(relation, start)
+        relation.where(arel_attribute(primary_key).gteq(bind_attribute(start)))
+      end
+
+      def apply_finish_limit(relation, finish)
+        relation.where(arel_attribute(primary_key).lteq(bind_attribute(finish)))
+      end
+
+      def bind_attribute(value)
+        predicate_builder.build_bind_attribute(primary_key, value)
```

Comment:
Using the predicate builder is fine to me. But the `bind_attribute` is too generic naming even though it is the primary key dedicated.

Reply:
Agreed. How about `primary_key_bind`? I amended the commit with that, but am open to other suggestions.